new-frontiers-14/frontier-station-14
# New admin command: spawnrefund

## About the PR
* Adds a new admin command: `spawnrefund`
* Adds a new admin log type: `AdminRefund`

## Why / Balance
Every once in a while, admins have to issue refunds. The server might crash, or a player sells someone else's ship, or a bug causes someone to be deleted, or raiders kill everyone on Frontier. Whatever the reason, refunds are sometimes needed. Unfortunately spawning exact amounts of spesos is *awful*. You have to spawn stacks of spesos and combine or split, which takes *ages*. Most admins seem to just round up to the nearest reasonable multiple so it doesn't take a decade. It's also a bit of a nightmare to ensure you don't drop money on the ground, as you can't spawn things directly into your own hand.

The `spawnrefund` command is intended to help with all of this.

Usage: `spawnrefund <amount> [reason]`
Examples:
* `spawnrefund 52000`
* `spawnrefund 52000 "for JeanGreytide, other player sold their ship"`

To limit abuse, the command imposes the following restrictions:
* You must be an admin with ADMIN privileges. TODO: Verify that this is the correct privilege.
* You must be a ghost. Stops potential badmins from spawning money whenever they need it. All our admins are of course beautiful people, but Frontier has downstreams, too. :)
* You must have a free hand. The command spawns the money directly into your hand.

The command logs every *successful* refund to the admin log. For ease of tracking, the message includes the word "refund", and there's a whole new log category called AdminRefund.

The command is called `spawnrefund` rather than something more generic, like `spawnspesos` or `spawnmoney`, because admins shouldn't really be spawning money for anything other than refunds.

## How to test
1. `deadmin` and try to use the command. Be rejected (and dejected).
2. `readmin` and possess a body. Use the command, be told you have to be a ghost.
3. `ghost` and use the command again. Be told you need a free hand.
4. `aghost` and use the command again. Get money in your hand.
5. Use the command again. Get money in your other hand. Verify no money drops on the ground.
6. Use the command again. Be told you need a free hand.
7. Try to `spawnrefund hello`. Be told `hello` is not a valid integer.
8. Try to `spawnrefund 0` and `spawnrefund -100`. Be told the values are out of range.

## Media
### Success
![image](https://github.com/user-attachments/assets/5bdf8b28-1f8b-49e9-9f62-6741f8fe87f9)

Admin log:
![image](https://github.com/user-attachments/assets/d9a97591-a0eb-4da4-a988-223df1067342)

Admin log message includes:
* The words "refund" and "spesos" for ease of finding without having to filter by log type.
* A reference to the spawned entity ID, so it can be tracked trivially, if necessary, without having to find a separate log message.

### Failure states
Non-admin attempt:
![image](https://github.com/user-attachments/assets/f00b09f9-ce00-4f75-8ee5-c6ae96eece56)

Admin, not attached to a ghost:
![image](https://github.com/user-attachments/assets/c1ffcc92-cdf4-40c9-9b8c-43c2ce8a4d96)

Non-admin ghost:
![image](https://github.com/user-attachments/assets/f6d2de38-6e94-4599-9182-0b2d31f28859)

No free hand:
![image](https://github.com/user-attachments/assets/fb14ad5d-7869-4cdb-9043-51ff3f3a3bd9)

Also shows a brief popup over the ghost in case you type the command into the chat console (`/spawnrefund ...`):
![image](https://github.com/user-attachments/assets/dfaed864-aacb-446f-ad53-d541d33e0a7e)

Argument validation errors:
![image](https://github.com/user-attachments/assets/088233c1-a28b-411d-b366-1abb4131a638)

## Requirements

## Breaking changes
None.

**Changelog**
:cl:
- add: There is a new admin command, `spawnrefund`, for spawning an exact number of spesos. To be used for the rare refund only, as the name suggests.
